```python
from enum import Enum, autoimport tracebackclass ExcHandler(Enum): RE_RAISE = auto() RAISE_NEW = auto() RAISE_FROM = auto() RAISE_FROM_NONE = auto() def print_reasons(ex): # see also https://docs.python.org/3/library/exceptions.html#exception-context print("The exception is of type: %s" % type(ex)) print("The cause from exception chaining is: %s" % type(ex.__cause__)) print("The context for exception chaining is: %s" % type(ex.__context__ )) print("Do suppress __context__ for exception chaining: %s" % ( ex.__suppress_context__) ) print("Traceback is the following:") print("-----------") traceback.print_exc() print("-----------") def divide(x, y, error_handling): try: result = x / y except Exception as ex: if error_handling == ExcHandler.RE_RAISE: raise elif error_handling == ExcHandler.RAISE_NEW: raise RuntimeError("raising new exception in except block") elif error_handling == ExcHandler.RAISE_FROM: raise RuntimeError("raising from original exception in except block") from ex elif error_handling == ExcHandler.RAISE_FROM_NONE: raise RuntimeError("raising from None in except block") from None# ------- main -------for stratagy in ExcHandler: print("=== Using exception handler: {} ===".format(stratagy)) input() try: divide(2, "invalid arg", stratagy) except Exception as ex: print_reasons(ex) print("======")
```
